Sources/Workspace/DefaultExtensionRunner.swift

```diff
@@ -0,0 +1,189 @@
+/*
+This source file is part of the Swift.org open source project
+
+Copyright (c) 2021 Apple Inc. and the Swift project authors
+Licensed under Apache License v2.0 with Runtime Library Exception
+
+See http://swift.org/LICENSE.txt for license information
+See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Basics
+import TSCBasic
+import TSCUtility
+import PackageGraph
+import PackageLoading   // for ManifestResourceProvider
+import PackageModel
+import SPMBuildCore
+import Foundation
+
+
+/// An extension runner that compiles the extension as an executable binary for the host platform, and invokes it as a subprocess.
+public struct DefaultExtensionRunner: ExtensionRunner {
+    
+    let cacheDir: AbsolutePath
+    let resources: ManifestResourceProvider
+
+    private static var _hostTriple = ThreadSafeBox<Triple>()
+    private static var _packageDescriptionMinimumDeploymentTarget = ThreadSafeBox<String>()
+    private let sdkRootCache = ThreadSafeBox<AbsolutePath>()
+
+    public init(cacheDir: AbsolutePath, manifestResources: ManifestResourceProvider) {
+        self.cacheDir = cacheDir
+        self.resources = manifestResources
+    }
+    
+    /// Public protocol function that compiles and runs the extension as a subprocess.  The tools version controls the availability of APIs in PackageExtension, and should be
+    public func runExtension(sources: Sources, inputJSON: Data, toolsVersion: ToolsVersion, diagnostics: DiagnosticsEngine, fileSystem: FileSystem) throws -> (outputJSON: Data, stdoutText: Data) {
+        let compiledExec = try self.compile(sources: sources, toolsVersion: toolsVersion, cacheDir: self.cacheDir)
+        return try self.invoke(compiledExec: compiledExec, input: inputJSON)
+    }
+
+    /// Helper function that compiles an extension as an executable and returns the path to it.
+    fileprivate func compile(sources: Sources, toolsVersion: ToolsVersion, cacheDir: AbsolutePath) throws -> AbsolutePath {
+        // FIXME: Much of this is copied from the ManifestLoader and should be consolidated.
```

Comment:
coud we move this to some utility helper that can be reused form both places? seem unfortunate to have this twice

Reply:
Yes, absolutely.  That's the consolidation work I mentioned, where this would be combined with that in ManifestLoader to have a single API for running scripts that are similar to PackageDescription or PackageExtension.  I'll make sure this is covered by a Radar and reference it here.